psi4/src/psi4/lib3index/dfhelper.cc

```diff
@@ -3107,6 +3223,71 @@ void DFHelper::compute_J(std::vector<SharedMatrix> D, std::vector<SharedMatrix>
         }
     }
 }
+void DFHelper::compute_J_combined(std::vector<SharedMatrix> D, std::vector<SharedMatrix> J, double* Mp, double* T1p, double* T2p, std::vector<std::vector<double>>& D_buffers, size_t bcount, size_t block_size) {
+//    double* coul_metp = metric_prep_core(1.0);
+    double ZERO[16] = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
+    double* zerop = &ZERO[0];
+    double* coul_metp;
+    std::unique_ptr<double[]> metric;
+
+
+    if (!hold_met_) {
+        metric = std::unique_ptr<double[]>(new double[naux_ * naux_]);
+        coul_metp = metric.get();
+        std::string filename = return_metfile(1.0);
+        get_tensor_(std::get<0>(files_[filename]), coul_metp, 0, naux_ - 1, 0, naux_ - 1);
+    } else
+        coul_metp = metric_prep_core(1.0);
+
+
+    for (size_t i = 0; i < J.size(); i++) {
+        double* Dp = D[i]->pointer()[0];
+        double* Jp = J[i]->pointer()[0];
+        C_DCOPY(naux_ *  nthreads_, zerop, 0, T1p, 1);
+        C_DCOPY(naux_, zerop, 0, T2p, 1);
+
+#pragma omp parallel for schedule(guided) num_threads(nthreads_)
+        for (size_t k = 0; k < nbf_; k++) {
+            size_t sp_size = small_skips_[k];
+            size_t jump = (AO_core_ ? big_skips_[k] + bcount * sp_size : (big_skips_[k] * block_size) / naux_ );
+            int rank = 0;
+#ifdef _OPENMP 
+            rank = omp_get_thread_num();
+#endif
+            for (size_t m = 0, sp_count = -1; m < nbf_; m++) {
+                if (schwarz_fun_mask_[k * nbf_ + m]) {
+                    sp_count++;
+                    D_buffers[rank][sp_count] = Dp[nbf_ * m + k];
+                }
+            }
+            C_DGEMV('N', block_size, sp_size, 1.0, &Mp[jump], sp_size, &D_buffers[rank][0], 1, 1.0, &T1p[rank * naux_], 1);
+        }
+        //reduce
+        for (size_t k = 1; k < nthreads_; k++) {
+            for (size_t l = 0; l < naux_; l++) { T1p[l] += T1p[ k * naux_ + l ]; }
+        }
+        //metric contraction
+        C_DGEMV('N', naux_, naux_, 1.0, coul_metp, naux_, T1p, 1, 0.0, T2p, 1);
+
+        C_DCOPY(nbf_ * nbf_, zerop, 0, T1p, 1);
```

Comment:
Same question here as on 3246 and 3247.

Reply:
memset looks like the right thing to do here as well